// skelet-core/lib/module.js
const freezeFunction = require('skelet-utils/lib/freezeFunction')

async function init (config) {
  await this.preInit(config)

  console.log(`${this.constructor.name}: init`)

  await this.postInit(config)

  this._loaded = true
}

async function unload (config) {
  await this.preUnload()

  Object.getOwnPropertyNames(this).forEach(property => delete this[property])

  Object.assign(this, new this.constructor())

  await this.postUnload()
}

async function reload (config) {
  await this.unload()
  await this._init(config)
}

/**
 *
 *  A basic module
 *
 *  exposes the following lifecycle methods:
 *  - init
 *  - unload
 *  - reload
 *
 *  exposes the following hooks
 *  - preInit
 *  - postInit
 *  - preUnload
 *  - postUnload
 */
class Module {
  constructor () {
    console.log(`${this.constructor.name}: constructor`)
    this._loaded = false

    freezeFunction(this, init)
    freezeFunction(this, unload)
    freezeFunction(this, reload)
  }

  get isLoaded () { return this._loaded }

  async preInit (config) {
    console.log(`${this.constructor.name}: preInit`)
  }

  async postInit (config) {
    console.log(`${this.constructor.name}: postInit`)
  }

  async preUnload () {
    console.log(`${this.constructor.name}: preUnload`)
  }

  async postUnload () {
    console.log(`${this.constructor.name}: postUnload`)
  }
}

module.exports = Module
